compiler/rustc_infer/src/infer/outlives/obligations.rs

```diff
@@ -186,6 +191,146 @@ impl<'cx, 'tcx> InferCtxt<'cx, 'tcx> {
             }
         }
     }
+
+    pub fn nested_outlives_obligations(
+        &self,
+        cause: &ObligationCause<'tcx>,
+        param_env: ty::ParamEnv<'tcx>,
+        ty: Ty<'tcx>,
+        region: ty::Region<'tcx>,
+    ) -> Vec<PredicateObligation<'tcx>> {
+        let to_obligation = |pred: ty::Binder<'tcx, ty::PredicateKind<'tcx>>| {
+            Obligation::new(cause.clone(), param_env, pred.to_predicate(self.tcx))
+        };
+        let ty_outlives = |ty| {
```

Comment:
maybe just make this eagerly recurse here? That should save a bit of back-and-forth between this function and the trait solver.